front-end/src/app/players/players.component.ts

```diff
@@ -0,0 +1,45 @@
+import { Component } from '@angular/core';
+import { Observable, map, tap } from 'rxjs';
+import { PlayerService } from '../player-service/player.service';
+import { GridColumn } from '../types/GridColumn';
+import { TeamService } from '../team-service/team.service';
+import { Player } from '../types/Player';
+import { Team } from '../types/Team';
+import { ClubService } from '../club-service/club.service';
+import { Club } from '../types/Club';
+
+interface PlayerRow extends Player {
+  club: string | undefined;
+  team: string | undefined;
+} 
+
+@Component({
+  selector: 'app-players',
+  templateUrl: './players.component.html',
+  styleUrls: ['./players.component.css']
+})
+export class PlayersComponent {
+  teams: Team[] = [];
+  clubs: Club[] = [];
+  public teamData$ = this.teamService.getTeams().subscribe(data => this.teams = data);
+  public clubData$ = this.clubService.getClubs().subscribe(data => this.clubs = data);
+  public rowData$: Observable<PlayerRow[]> = this.playerService.getPlayers().pipe(
+    map(players => players.map(player => {
+      const team = this.teams.find(team => team.id === player.teamId);
+      const club = this.clubs.find(club => club.id === team?.clubId);
```

Comment:
For every player (not just ones that will be visible on the grid), we are searching through every team and club. When you get the teams and the clubs above, you could store them in a hash-map from teamId to the team name (and clubId to club name). Then, you don't need to go through the full teams and clubs lists for each player. Does that make sense?